storage/src/main/java/io/delta/storage/GCSLogStore.java

```diff
@@ -0,0 +1,189 @@
+    }
+
+    private static <T> T runInNewThread(
+            String threadName,
+            Boolean isDaemon,
+            Callable<T> body) throws Exception {
+        List<Exception> exceptionHolder = new ArrayList<>(1);
+        List<T> resultHolder = new ArrayList<>(1);
+        Thread thread = new Thread(threadName) {
+            @Override
+            public void run() {
+                try {
+                    resultHolder.add(body.call());
+                } catch (Exception ex) {
+                    exceptionHolder.add(ex);
+                }
+            }
+        };
+        thread.setDaemon(isDaemon);
+        thread.start();
+        thread.join();
+        if (!exceptionHolder.isEmpty()) {
+            Exception realException = exceptionHolder.get(0);
+            // Remove the part of the stack that shows method calls into this helper method
+            // This means drop everything from the top until the stack element
+            // ThreadUtils.runInNewThread(), and then drop that as well (hence the `drop(1)`).
+            Stream<StackTraceElement> baseStackTrace = Arrays.stream(
+                    Thread.currentThread().getStackTrace()).dropWhile(t ->
+                    !t.getClassName().contains(GCSLogStore.class.getSimpleName())).skip(1);
+            // Remove the part of the new thread stack that shows methods call from this helper method
+            Stream<StackTraceElement> extraStackTrace = Arrays.stream(
+                    realException.getStackTrace()).takeWhile(e ->
+                    !e.getClassName().contains(GCSLogStore.class.getSimpleName()));
+            // Combine the two stack traces, with a place holder just specifying that there
+            // was a helper method used, without any further details of the helper
+            Stream<StackTraceElement> placeHolderStackElem = Stream.of(
+                    new StackTraceElement(
+                            String.format(
+                                    "... run in separate thread using $s",
+                                    GCSLogStore.class.getSimpleName(),
+                                    " static method runInNewThread"),
+                            "",
+                            "",
+                            -1)
+            );
+            StackTraceElement[] finalStackTrace = Stream.concat(
+                    Stream.concat(baseStackTrace,
+                            placeHolderStackElem),
+                    baseStackTrace).toArray(StackTraceElement[]::new);
+            realException.setStackTrace(finalStackTrace);
+            throw realException;
+        } else {
+            return resultHolder.get(0);
+        }
+    }
+
+    @Override
+    public void write(Path path,
+                      Iterator<String> actions,
+                      Boolean overwrite,
+                      Configuration hadoopConf) throws IOException {
+        FileSystem fs = path.getFileSystem(hadoopConf);
```

Comment:
in general, try and declare variables as `final` when you can.

Reply:
Will keep in mind